src/components/AppNavigator/AppNavigator.tsx

```diff
@@ -1,100 +1,111 @@
-      })
-      .map(([screen, step]) => {
-        const { isCompleted } = step;
-        const url = routes[screen];
-
-        return { isCompleted, step: screen, url };
-      })
-      .reduce<MenuItemType[]>((accumulator, element) => {
-        const { step } = element;
-
-        if (
-          step === FormScreens.DELIVERY_MODE ||
-          step === FormScreens.PAYMENT_METHOD
-        ) {
-          const subStep = formValues[step]?.subStep;
-
-          if (subStep) {
-            const menuItem = {
-              isCompleted: false,
-              step: subStep.id,
-              url: routes[subStep.id],
-            };
-            accumulator.push(element, menuItem);
-          } else {
-            accumulator.push(element);
-          }
-        } else {
-          accumulator.push(element);
+  useEffect(() => {
+    const formScreen = (Object.keys(routes) as (keyof typeof routes)[]).find(
+      (key) => routes[key] === pathname,
+    );
+    const findStep = () => {
+      // eslint-disable-next-line no-restricted-syntax
+      for (const [screenName, step] of formValuesEntries) {
+        if (screenName === formScreen) {
+          return step;
         }
+        if (step.subStep) {
+          return step.subStep;
+        }
+      }
+      throw new Error('Step/substep not found');
+    };
 
-        return accumulator;
-      }, []);
-
-    return menuItemsList;
-  };
-
-  const menuItemsList = getMenuItemsFromSteps(formValues);
-
-  const getActiveMenuItem = (path: string): number => {
-    const index = menuItemsList.findIndex((step) => step.url === path);
+    const currentStep = findStep();
+    if (!currentStep.isCompleted && formScreen !== firstUncompletedStep) {
+      navigate(routes[firstUncompletedStep]);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [firstUncompletedStep, pathname, navigate]);
 
-    if (index === -1) {
-      throw new Error('Menu item not found');
+  const getActiveStepNumber = () => {
+    let result = 0;
+    // eslint-disable-next-line no-restricted-syntax
+    for (const step of Object.values<FormStepsList>(formValues)) {
+      if (step.isCompleted) {
+        result += 1;
+      }
     }
 
-    return index;
+    return result;
   };
 
+  const renderMenuItem = (screenName: FormScreens, step: FormStepsList) => (
+    <Step key={screenName}>
+      <StepLabel>
+        {step.isCompleted || screenName === firstUncompletedStep ? (
+          <Link
+            component={RouterLink}
+            to={routes[screenName]}
+            underline='hover'
+            variant='overline'>
+            {screenName}
+          </Link>
+        ) : (
+          <Typography
+            component='span'
+            variant='overline'>
+            {screenName}
+          </Typography>
+        )}
+      </StepLabel>
+    </Step>
+  );
+
+  const renderMenuItemList = formValuesEntries.reduce<JSX.Element[]>(
+    (prev, [screenName, step]) =>
+      step.subStep
```

Comment:
Good `.reduce` signature.
In order to fix TS issue, we can try using type guard that checks that current step isn't substep